delta/plugins/storage/src/main/scala/ch/epfl/bluebrain/nexus/delta/plugins/storage/storages/operations/s3/S3StorageSaveFile.scala

```diff
@@ -1,76 +1,169 @@
+
+  def apply(
       filename: String,
       entity: BodyPartEntity
-  ): IO[FileStorageMetadata] = {
+  ): IO[FileStorageMetadata] =
     for {
       uuid   <- uuidf()
       path    = Uri.Path(intermediateFolders(storage.project, uuid, filename))
       result <- storeFile(path, uuid, entity)
     } yield result
-  }
 
   private def storeFile(path: Path, uuid: UUID, entity: BodyPartEntity): IO[FileStorageMetadata] = {
-    val key        = path.toString()
-    val attributes = S3Attributes.settings(storage.value.alpakkaSettings(config))
-    def s3Sink     = S3.multipartUpload(storage.value.bucket, key).withAttributes(attributes)
-    IO.fromFuture(
-      IO.delay(
-        S3.getObjectMetadata(storage.value.bucket, key)
-          .withAttributes(attributes)
-          .runWith(Sink.last)
-          .flatMap {
-            case None    =>
-              entity.dataBytes.runWith(SinkUtils.combineMat(digestSink(storage.value.algorithm), sizeSink, s3Sink) {
-                case (digest, bytes, s3Result) =>
-                  Future.successful(
-                    FileStorageMetadata(
-                      uuid = uuid,
-                      bytes = bytes,
-                      digest = digest,
-                      origin = Client,
-                      location = s3Result.location.withPath(Slash(path)),
-                      path = Uri.Path(key)
-                    )
-                  )
-              })
-            case Some(_) => Future.failed(fileAlreadyExistException)
-          }
+    val key                        = path.toString()
+    val fileData: Stream[IO, Byte] = convertStream(entity.dataBytes)
+
+    (for {
+      _          <- log(key, s"Checking for object existence")
+      _          <- validateObjectDoesNotExist(key)
+      _          <- log(key, s"Beginning multipart upload")
+      maybeEtags <- uploadFileMultipart(fileData, key)
+      _          <- log(key, s"Finished multipart upload. Etag by part: $maybeEtags")
+      attr       <- collectFileMetadata(fileData, key, uuid, maybeEtags)
+    } yield attr)
+      .onError(e => logger.error(e)("Unexpected error when storing file"))
+      .adaptError { err => UnexpectedSaveError(key, err.getMessage) }
+  }
+
+  private def validateObjectDoesNotExist(key: String) =
+    getFileAttributes(key).redeemWith(
+      {
+        case _: NoSuchKeyException => IO.unit
+        case e                     => IO.raiseError(e)
+      },
+      _ => IO.raiseError(ResourceAlreadyExists(key))
+    )
+
+  private def convertStream(source: Source[ByteString, Any]): Stream[IO, Byte] =
+    StreamConverter(
+      source
+        .flatMapConcat(x => Source.fromIterator(() => x.iterator))
+        .mapMaterializedValue(_ => NotUsed)
+    )
+
+  private def uploadFileMultipart(fileData: Stream[IO, Byte], key: String): IO[List[Option[ETag]]] =
+    fileData
+      .through(
+        s3.uploadFileMultipart(
```

Comment:
We could do a put object here, not sure what the multipart brings here.
Is there only 1 part in the end ?